Rishavganguly007/feature/api football api integration #5646

rishavganguly007 opened the pull request:
## Description

This PR adds a [Football API](https://www.api-football.com/) handler to get player information.
Tested extensively locally.
**Fixes** #(5646)

## Type of change

- [x] ⚡ New feature (non-breaking change which adds functionality)

lnxpy commented on mindsdb/integrations/handlers/footballApi_handler/README.md:
````diff
@@ -0,0 +1,37 @@
+We start by creating a database to connect to the Football API. You'll need an access token which can be accessed from [RAPIDAPI](https://api-football-v1.p.rapidapi.com/v3/) or  [API-SPORTS](https://v3.football.api-sports.io/) and the domain you want to send API requests to.  
+  
+Example  
+```  
````
Would you add the `sql` code-block highlighter?

```
```sql
...
```

rishavganguly007 replied:
fixed in commit 02e3d0433031ea39532fe3f3f93e498102429c2e

lnxpy commented on mindsdb/integrations/handlers/footballApi_handler/README.md:
````diff
@@ -0,0 +1,37 @@
+
+  
+
+```
````
I guess it's misplaces.

rishavganguly007 replied:
fixed in commit 02e3d0433031ea39532fe3f3f93e498102429c2e

lnxpy commented on mindsdb/integrations/handlers/footballApi_handler/README.md:
````diff
@@ -0,0 +1,37 @@
+;## Get Player Info 
+To get the player information, use the following query 
+  
+```  
````
Same `sql` highlighter here.

rishavganguly007 replied:
fixed in commit 02e3d0433031ea39532fe3f3f93e498102429c2e

lnxpy reviewed (changes requested):
Also, please change the naming convention that you've used during your integration. Some names that you've used for your entities and files can be corrected and _pep-proved_.

I highly recommend you install [`pre-commit`](https://pre-commit.com/) and run the hooks over on your changes.

```sh
pre-commit run --files path/to/integration/*.py --show-diff-on-failure
```

P.S. Just so you know – these change requests have nothing to do with the performance of your integration. It just guarantees maintainability. Thank you for your lovely contribution. :)

lnxpy commented:
@rishavganguly007 Would you squash your commits into one?

rishavganguly007 commented:
Can you tell how can I do that?

lnxpy commented:
@rishavganguly007 Sure. You've done six commits.

```sh
git rebase -i HEAD~6
```

Then. change the word `pick` to `squash` from the beginning of the last 5 lines. Save the prompt and set a new git commit.

Afterward, push your changes with `-f`.

lnxpy commented:
@rishavganguly007 Would you share some demo photos/clips of the integration? (you can upload it within the PR comment)

rishavganguly007 commented:
> @rishavganguly007 Would you squash your commits into one?
Done.

> @rishavganguly007 Would you share some demo photos/clips of the integration? (you can upload it within the PR comment)
PFA snapshots and csv file of the data

Creating the database 
![creat_api-footbal_db](https://github.com/mindsdb/mindsdb/assets/79499308/dd9cc1c4-d303-4db5-a3dc-3db98b5054bb)

Fetching the Player data
![getPlayer_api-football_db](https://github.com/mindsdb/mindsdb/assets/79499308/07182a3b-6bd6-48b7-9215-3cddfa67c3ce)

player data
[API-Football_player_data.csv](https://github.com/mindsdb/mindsdb/files/12641049/API-Football_player_data.csv)

lnxpy commented:
Nice!
What do you think about this?! @ZoranPandovski

ZoranPandovski commented on mindsdb/integrations/handlers/footballApi_handler/FootballApi_Tables.py:
```diff
@@ -0,0 +1,150 @@
+                """
+        conditions = extract_comparison_conditions(query.where)
+        player_params = {}
+        for a_where in conditions:
```
Can we simplify this as:

```
 key = a_where[1]
    
    if a_where[0] != "=":
        raise ValueError(f"Unsupported where operation for {key}")
    
    # Assign values based on the key
    if key == "league":
        player_params["league"] = int(a_where[2])
    elif key == "season":
        player_params["season"] = int(a_where[2])
    elif key == "team":
        player_params["team"] = int(a_where[2])
    elif key == "id":
        player_params["id"] = int(a_where[2])
    elif key == "search":
        player_params["search"] = a_where[2]
    elif key == "page":
        player_params["page"] = int(a_where[2])
    else:
        raise ValueError(f"Unsupported where argument {key}")
```

ZoranPandovski commented on mindsdb/integrations/handlers/footballApi_handler/README.md:
````diff
@@ -0,0 +1,35 @@
+```sql
+CREATE  DATABASE my_mindsdb_footballapi
+
+WITH
````
Please format the SQL query

ZoranPandovski commented on mindsdb/integrations/handlers/footballApi_handler/README.md:
````diff
@@ -0,0 +1,35 @@
+};
+```  
+  
+;## Get Player Info 
````
Fix the title, leave only ##

ZoranPandovski commented on mindsdb/integrations/handlers/footballApi_handler/README.md:
````diff
@@ -0,0 +1,35 @@
+FROM my_mindsdb_footballapi.get_players
+WHERE league=39  AND season=2021 AND page=3;  
+```  
+-  id = id of the player
````
Maybe remove this list

ZoranPandovski commented on mindsdb/integrations/handlers/footballApi_handler/footballApi_handler.py:
```diff
@@ -0,0 +1,175 @@
+
+        return player_data_list
+
+    def _flatten_json(self, json_data, prefix=''):
```
Please add comment on what this function does

ZoranPandovski reviewed (changes requested):
Looks great, please check the few comments I've left

rishavganguly007 commented:
> Looks great, please check the few comments I've left

Hi @ZoranPandovski , I have addressed the comments that you had left, please let me know if anything else I have to do.

MinuraPunchihewa commented on mindsdb/integrations/handlers/footballApi_handler/footballApi_handler.py:
```diff
@@ -0,0 +1,186 @@
+        ast = parse_sql(query, dialect='mindsdb')
+        return self.query(ast)
+
+    def call_football_api(self, method_name: str, **player_params) -> pd.DataFrame:
```
If we do create a `PlayersTable` as I have suggested above, I think we should probably move these functions that seem to be specific to extracting player data, to that class. The `call_football_api()` function does not seem to be generic per se, since it takes in the name of a method and then just calls it. I would think that it'd be much simple to call the relevant method inside of each class that we create: `PlayersTable`, `TeamsTable` etc.

ZoranPandovski commented:
@rishavganguly007 Can you please address the latest changes required from @MinuraPunchihewa

rishavganguly007 commented:
> @rishavganguly007 Can you please address the latest changes required from @MinuraPunchihewa

Hi @ZoranPandovski , @MinuraPunchihewa 

I have added a latest commit addressing the changes mentioned by @MinuraPunchihewa , please review and let me know .

Thanks

MinuraPunchihewa commented on mindsdb/integrations/handlers/footballApi_handler/footballApi_handler.py:
```diff
@@ -0,0 +1,122 @@
+        self.connection_data = connection_data
+        self.connection = None
+        players = PlayersTable(self)
+        self._register_table("get_players", players)
```
I think we should change the name of the table to `players`. We are going for the concept of tables (API Tables) here, so `get_players` does not make much sense. If you do agree with this, you will need to change the examples in the README as well.

MinuraPunchihewa commented on mindsdb/integrations/handlers/footballApi_handler/players_table.py:
```diff
@@ -0,0 +1,134 @@
+            else:
+                raise ValueError(f"Unsupported where argument {key}")
+
+        player_data = self.handler.call_football_api("get_players", **player_params)
```
I do not really understand the need to pass the method name to the `call_football_api()` function in the handler and then call that method from there. I think you can directly call the methods of the `PlayersApi` class here. It is fine to use the `call_football_api()` function, granted that you have a somewhat generic approach to call the Football API for any resource (players, teams etc.) and that it reduces the amount of code required for the implementation, but this does not really seem to do that. The use of the config (`footballapi_constants.py`) does not seem to be necessary either.

rishavganguly007 commented:
Hi 

Let me know if the PR is good to merge or we need some more changes

Thanks

MinuraPunchihewa commented:
Hey @rishavganguly007,
Can you address the last comment I have made? After that, I think we can merge this.

rishavganguly007 commented:
> Hey @rishavganguly007, Can you address the last comment I have made? After that, I think we can merge this.

Hi, I have removed the `call_football_api` function. Let me know if the PR can be merged or any further changes is required.

MinuraPunchihewa commented on mindsdb/integrations/handlers/footballApi_handler/__about__.py:
```diff
@@ -0,0 +1,9 @@
+__author__ = "Rishav Ganguly"
+__github__ = "https://github.com/mindsdb/mindsdb"
+__pypi__ = "https://pypi.org/project/mindsdb/"
+__license__ = "GPL-3.0"
```
Can we change the licence to `MIT`, please? We have made this update recently.

MinuraPunchihewa commented on mindsdb/integrations/handlers/footballApi_handler/footballApi_handler.py:
```diff
@@ -0,0 +1,98 @@
+        connection_data = kwargs.get('connection_data', {})
+        if 'api_key' in connection_data:
+            self.api_key = connection_data['api_key']
+        elif 'API_KEY' in os.environ:
```
Since we have many integrations that use API keys, I think we need to be a little more specific on the name of the environment variable used. In order to extract the API key, can you please use `get_api_key()` method from `mindsdb.integrations.utilities.handler_utils`? This has been created for this purpose.

MinuraPunchihewa commented on mindsdb/integrations/handlers/footballApi_handler/players_table.py:
```diff
@@ -0,0 +1,138 @@
+            "player_weight",
+            "player_injured",
+            "player_photo",
+            "statistics_0_team_id",
```
Is it possible to remove the `0` in these column names? This is not a big problem, but I was just wondering if it is possible because from what I could see the API only returns one record for each of these statistics?

MinuraPunchihewa reviewed (commented):
Hi @rishavganguly007,
I've just left a couple more comments for small improvements we could (or need to) make.

If you do have the time, please make the following naming changes (not mandatory),
`footballApi_handler` (package) -> `fotballapi_handler`
`footballApi_handler` (module) -> `fotballapi_handler`
`FootballApiHandler` (class) -> `FootballAPIHandler`
`PlayersApi` (class) -> `PlayersAPI`

ZoranPandovski commented:
Closing as outdated
